src/core/ports/ark_url_formatter.rs

```diff
@@ -0,0 +1,304 @@
+impl<T: ArkUrlFormatterPort> ArkUrlFormatterPortExt for T {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::core::errors::ark_url_formatter::ArkUrlFormatterError;
+
+    // Mock implementation for testing
+    struct MockArkUrlFormatter;
+
+    impl ArkUrlFormatterPort for MockArkUrlFormatter {
+        fn resource_iri_to_ark_id(
+            &self,
+            resource_iri: &str,
+            timestamp: Option<&str>,
+        ) -> ArkUrlFormatterResult<String> {
+            if resource_iri.is_empty() {
+                return Err(ArkUrlFormatterError::InvalidResourceIri(
+                    resource_iri.to_string(),
+                ));
+            }
+
+            let mut result = format!("ark:/00000/1/0001/{}", resource_iri);
+            if let Some(ts) = timestamp {
+                result.push('.');
+                result.push_str(ts);
+            }
+            Ok(result)
+        }
+
+        fn resource_iri_to_ark_url(
+            &self,
+            resource_iri: &str,
+            value_id: Option<&str>,
+            timestamp: Option<&str>,
+        ) -> ArkUrlFormatterResult<String> {
+            if resource_iri.is_empty() {
+                return Err(ArkUrlFormatterError::InvalidResourceIri(
+                    resource_iri.to_string(),
+                ));
+            }
+
+            let mut result = format!("https://ark.example.org/ark:/00000/1/0001/{}", resource_iri);
+
+            if let Some(vid) = value_id {
+                result.push('/');
+                result.push_str(vid);
+            }
+
+            if let Some(ts) = timestamp {
+                result.push('.');
+                result.push_str(ts);
+            }
+
+            Ok(result)
+        }
+
+        fn format_ark_url(
+            &self,
+            project_id: &str,
+            resource_id_with_check_digit: &str,
+            value_id_with_check_digit: Option<&str>,
+            timestamp: Option<&str>,
+        ) -> ArkUrlFormatterResult<String> {
+            if project_id.is_empty() {
+                return Err(ArkUrlFormatterError::InvalidProjectId(
+                    project_id.to_string(),
+                ));
+            }
+
+            let mut result = format!(
+                "https://ark.example.org/ark:/00000/1/{}/{}",
+                project_id, resource_id_with_check_digit
+            );
+
+            if let Some(vid) = value_id_with_check_digit {
+                result.push('/');
+                result.push_str(vid);
+            }
+
+            if let Some(ts) = timestamp {
+                result.push('.');
+                result.push_str(ts);
+            }
+
+            Ok(result)
+        }
+    }
+
+    #[test]
+    fn test_port_resource_iri_to_ark_id() {
+        let formatter = MockArkUrlFormatter;
+        let result = formatter.resource_iri_to_ark_id("test", None).unwrap();
+        assert_eq!(result, "ark:/00000/1/0001/test");
+    }
```

Comment:
I haven't gone through all these tests in detail, but I think there may be more cases.  
Looking at this test, all it does is testing the mock implementation - which I don't think is a useful test. And I don't really see how you could test a port (that is, an interface) in a meaningful way, because by it's very nature, interfaces don't have logic, only type signature, and so there is no logic to test there.